File: hnaj-be/app/Exceptions/AuthFlowException.php
<?php

namespace App\Exceptions;

use App\Enums\AuthErrorCode;
use RuntimeException;

/**
 * Lỗi nghiệp vụ của luồng authentication.
 * Mang sẵn HTTP status và mã lỗi để handler dựng error envelope chung.
 */
class AuthFlowException extends RuntimeException
{
    public function __construct(
        public readonly AuthErrorCode $errorCode,
        string $message,
        public readonly int $status = 422,
    ) {
        parent::__construct($message);
    }

    public static function invalidCredentials(): self
    {
        return new self(
            AuthErrorCode::InvalidCredentials,
            'The provided credentials are incorrect.',
            401,
        );
    }

    public static function emailNotVerified(): self
    {
        return new self(
            AuthErrorCode::EmailNotVerified,
            'This email address has not been verified yet.',
            403,
        );
    }

    public static function accountNotActive(): self
    {
        return new self(
            AuthErrorCode::AccountNotActive,
            'This account is not active.',
            403,
        );
    }

    public static function invalidVerificationToken(): self
    {
        return new self(
            AuthErrorCode::InvalidToken,
            'This verification link is invalid or has expired.',
            422,
        );
    }

    public static function emailAlreadyVerified(): self
    {
        return new self(
            AuthErrorCode::AlreadyVerified,
            'This email address has already been verified.',
            409,
        );
    }

    public static function forbiddenRole(): self
    {
        return new self(
            AuthErrorCode::ForbiddenRole,
            'You do not have permission to access this resource.',
            403,
        );
    }

    public static function googleAuthFailed(string $message = 'Unable to complete Google sign-in.'): self
    {
        return new self(AuthErrorCode::GoogleAuthFailed, $message, 422);
    }
}

File: hnaj-be-java/src/main/java/com/hnaj/api/exception/AuthFlowException.java
package com.hnaj.api.exception;

/**
 * Lỗi nghiệp vụ luồng authentication. Mang sẵn HTTP status + error code
 * để ApiExceptionHandler dựng envelope chung.
 */
public class AuthFlowException extends RuntimeException {

    private final AuthErrorCode errorCode;
    private final int status;

    public AuthFlowException(AuthErrorCode errorCode, String message, int status) {
        super(message);
        this.errorCode = errorCode;
        this.status = status;
    }

    public AuthErrorCode getErrorCode() {
        return errorCode;
    }

    public int getStatus() {
        return status;
    }

    // -------- Factory methods mapping Laravel AuthFlowException --------

    public static AuthFlowException invalidCredentials() {
        return new AuthFlowException(AuthErrorCode.INVALID_CREDENTIALS,
                "The provided credentials are incorrect.", 401);
    }

    public static AuthFlowException emailNotVerified() {
        return new AuthFlowException(AuthErrorCode.EMAIL_NOT_VERIFIED,
                "This email address has not been verified yet.", 403);
    }

    public static AuthFlowException accountNotActive() {
        return new AuthFlowException(AuthErrorCode.ACCOUNT_NOT_ACTIVE,
                "This account is not active.", 403);
    }

    public static AuthFlowException invalidVerificationToken() {
        return new AuthFlowException(AuthErrorCode.INVALID_VERIFICATION_TOKEN,
                "This verification link is invalid or has expired.", 422);
    }

    public static AuthFlowException emailAlreadyVerified() {
        return new AuthFlowException(AuthErrorCode.EMAIL_ALREADY_VERIFIED,
                "This email address has already been verified.", 409);
    }

    public static AuthFlowException forbiddenRole() {
        return new AuthFlowException(AuthErrorCode.FORBIDDEN_ROLE,
                "You do not have permission to access this resource.", 403);
    }

    public static AuthFlowException googleAuthFailed(String message) {
        return new AuthFlowException(AuthErrorCode.GOOGLE_AUTH_FAILED, message, 422);
    }

    public static AuthFlowException googleAuthFailed() {
        return googleAuthFailed("Unable to complete Google sign-in.");
    }
}
